File: backend/app/schemas/anomaly.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import NdviStatus


class AnomalyOut(BaseModel):
    """Период аномалии, как его ждёт Flutter-клиент (`frontend/lib/models/anomaly.dart`)."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., examples=["AOI-0001-2025-06-14"])
    anon_polygon_id: str = Field(..., validation_alias="polygon_id", examples=["AOI-0001"])
    start_date: date = Field(..., examples=["2025-06-14"])
    end_date: date = Field(..., examples=["2025-09-01"])
    severity: NdviStatus = Field(
        ..., description="'suppression' (угнетение биомассы) или 'critical' (критическая аномалия)",
        examples=["critical"],
    )
    min_z_score: float = Field(..., description="Наихудший Z-score в периоде", examples=[-2.4])
    deviation: float = Field(
        ..., description="Отклонение NDVI от климатической нормы (отрицательное значение)", examples=[-0.31]
    )
    explanation: str = Field(
        "",
        description="Текстовая интерпретация вероятной причины аномалии (сопоставление с ERA5)",
        examples=[
            "NDVI ниже климатической нормы на 0.31; осадков за период почти не было "
            "(2.1 мм) — вероятна почвенная засуха."
        ],
    )

File: frontend/lib/models/anomaly.dart
import 'ndvi_point.dart';

/// Период аномалии образуют только два «не штатных» статуса [NdviStatus].
typedef AnomalySeverity = NdviStatus;

/// Непрерывный отрезок дат, на которых Z-score опустился ниже -1
/// (угнетение) или -2 (критическая аномалия) — см. [ndviStatusForZ].
/// [severity] — худший статус, достигнутый где-либо внутри отрезка.
class Anomaly {
  final String id;
  final String polygonId;
  final DateTime startDate;
  final DateTime endDate;
  final AnomalySeverity severity;
  final double minZScore;
  /// Отклонение от климатической нормы, например -0.35 значит, что NDVI
  /// упал на 0.35 ниже ожидаемого сезонного значения.
  final double deviation;
  final String explanation;

  const Anomaly({
    required this.id,
    required this.polygonId,
    required this.startDate,
    required this.endDate,
    required this.severity,
    required this.minZScore,
    required this.deviation,
    required this.explanation,
  });

  factory Anomaly.fromJson(Map<String, dynamic> json) {
    return Anomaly(
      id: json['id'] as String,
      polygonId: json['anon_polygon_id'] as String,
      startDate: DateTime.parse(json['start_date'] as String),
      endDate: DateTime.parse(json['end_date'] as String),
      severity: NdviStatus.values.firstWhere(
        (s) => s.name == json['severity'],
        orElse: () => NdviStatus.suppression,
      ),
      minZScore: (json['min_z_score'] as num).toDouble(),
      deviation: (json['deviation'] as num).toDouble(),
      explanation: json['explanation'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'anon_polygon_id': polygonId,
        'start_date': startDate.toIso8601String(),
        'end_date': endDate.toIso8601String(),
        'severity': severity.name,
        'min_z_score': minZScore,
        'deviation': deviation,
        'explanation': explanation,
      };
}
